processor/schemaprocessor/internal/migrate/attributes.go

```diff
@@ -0,0 +1,108 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migrate // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/migrate"
+
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/alias"
+)
+
+// AttributeKey is used to allow strings and type defined
+// strings to be used within the creation of creating an `AttributeChangeSet`
+type AttributeKey interface {
+	~string
+}
+
+// AttributeChangeSet represents an unscoped entry that can be applied.
+//
+// The listed changes are duplicated twice
+// to allow for simplified means of transition to or from a revision.
+type AttributeChangeSet struct {
+	updates  map[alias.AttributeKey]alias.AttributeKey
+	rollback map[alias.AttributeKey]alias.AttributeKey
+}
+
+// OrderedAttributeChangeSets allows for `AttributeChangeSet`
+// to be chained together as they are defined within the schema
+// and be applied sequentially to ensure deterministic behavior.
+type AttributeChangeSetSlice []*AttributeChangeSet
+
+// NewAttributes allows for typed strings to be used as part
+// of the invocation that will be converted into the default string type.
+func NewAttributes[Key AttributeKey, Value AttributeKey](mappings map[Key]Value) *AttributeChangeSet {
+	attr := &AttributeChangeSet{
+		updates:  make(map[string]string, len(mappings)),
+		rollback: make(map[string]string, len(mappings)),
+	}
+	for k, v := range mappings {
+		attr.updates[string(k)] = string(v)
+		attr.rollback[string(v)] = string(k)
+	}
+	return attr
+}
+
+func (a *AttributeChangeSet) Apply(attrs pcommon.Map) {
+	a.do(StateSelctorApply, attrs)
+}
+
+func (a *AttributeChangeSet) Rollback(attrs pcommon.Map) {
+	a.do(StateSelectorRollback, attrs)
+}
+
+func (a *AttributeChangeSet) do(ss StateSelctor, attrs pcommon.Map) {
+	attrs.RemoveIf(func(s string, v pcommon.Value) (matched bool) {
+		var key string
+		switch ss {
+		case StateSelctorApply:
+			key, matched = a.updates[s]
+		case StateSelectorRollback:
+			key, matched = a.rollback[s]
+		}
+		if matched {
+			v.CopyTo(attrs.PutEmpty(key))
```

Comment:
Can we add a TODO to decide what we should do if the key already exists? Should we overwrite it (as it does now) or we should we stop and return an error? This is probably more of a TODO for schema spec.

Reply:
The `RemoveIf` contract does not say whether it is ok to modify the attrs from the callback. Can we confirm that this is OK and if so make sure `RemoveIf` comments explicitly bless it?

Reply:
Let me follow this one up, because it works and doesn't cause a race condition according to the race detector.

Reply:
Alright, apparently it isn't recommend so I have updated the actual work that it is doing.

Reply:
> Can we add a TODO to decide what we should do if the key already exists? Should we overwrite it (as it does now) or we should we stop and return an error? This is probably more of a TODO for schema spec.

I've made it so it will prefer the schema rewrite values over the existing values to ensure it is deterministic. Meaning that values that are not part of the schema will be removed.

Reply:
@mx-psi yes, sounds good to me.